### PR TITLE
Optimizations for MXFP8/NVFP4 dequantize kernels

### PR DESCRIPTION
**Pull request by YigongQin**
# Description

  - Handle empty tensors in dequantize for CUDA graph compatibility
  - Add swizzled scale support to the NVFP4 dequantize kernel, reusing the existing MXFP8 swizzle index computation
  - Add C++ unit tests for both NVFP4 and MXFP8 dequantization (including swizzled scale variants)
  - Fix to_cpu() and set_scale() in test infrastructure to correctly sync amax/scale for NVTE_NVFP4_1D_SCALING mode

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

-   Handle empty tensors in dequantize for CUDA graph compatibility — Early return when input has zero elements, avoiding kernel launches on empty tensors.
-   Add GEMM-swizzled scale support to NVFP4 dequantize kernel — Template the kernel with WITH_GEMM_SWIZZLED_SCALES to support reading scales from swizzled layout, reusing the MXFP8 swizzle index computation.
-   Add GEMM-swizzled scale support to MXFP8 dequantize kernel — Extend the MXFP8 dequantize kernel to handle swizzled scale inputs.
-   Add C++ unit tests for NVFP4 dequantization — 21 tests for compact scales + 21 tests for swizzled scales, covering multiple sizes and output dtypes (fp32, bf16, fp16).
-   Add C++ unit tests for MXFP8 dequantization with swizzled scales — New swizzled test suite for MXFP8.
-   Fix to_cpu() to sync amax/scale for NVFP4 tensors — Previously only synced for NVTE_DELAYED_TENSOR_SCALING, causing the CPU reference to use stale amax=0.
-   Fix set_scale() to work for NVFP4 tensors — Same condition fix, enabling the scale to be properly uploaded to GPU before quantization.
-   Fix swizzled test ordering — Move from_cpu() before the FP4 data copy to prevent from_cpu() from overwriting the copied data with zeros.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by zianglih** [comment]
The following relevant unit tests passed on SM100 (with the drop `optimize_for_gemm = False` changes):
```
python3 -m pytest --tb=auto tests/pytorch/test_backward_override.py
python3 -m pytest --tb=auto tests/pytorch/test_sanity.py
python3 -m pytest --tb=auto tests/pytorch/test_cpu_offloading.py
PYTORCH_JIT=0 NVTE_TORCH_COMPILE=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 NVTE_FUSED_ATTN=0 python3 -m pytest --tb=auto tests/pytorch/test_cuda_graphs.py
NVTE_TEST_NVINSPECT_ENABLED=1 NVTE_TEST_NVINSPECT_CONFIG_FILE=tests/pytorch/debug/test_configs/dummy_feature.yaml NVTE_TEST_NVINSPECT_FEATURE_DIRS=transformer_engine/debug/features PYTORCH_JIT=0 NVTE_TORCH_COMPILE=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 python3 -m pytest --tb=auto tests/pytorch/test_sanity.py
```

**Comment by zianglih** [comment]
After this PR, fwd is around 3%-4% faster for DeepSeek shape MoE:
```
# With the optimization
NVTE_BACKWARD_OVERRIDE=dequantized python benchmarks/linear/benchmark_grouped_linear.py --recipe mxfp8 --fwd-only
       m     k     n recipe  num_gemms  grouped_fwd_time_ms
0  16384  7168  2048  mxfp8          4             0.272829
1  32768  7168  2048  mxfp8          4             0.509788
2  65536  7168  2048  mxfp8          4             0.948633
3  98304  7168  2048  mxfp8          4             1.391146
0  16384  7168  2048  mxfp8          8             0.303238
1  32768  7168  2048  mxfp8          8             0.533896
2  65536  7168  2048  mxfp8          8             1.003446
3  98304  7168  2048  mxfp8          8             1.470030

# Without the optimization
git restore --source 77b8681de5cf -- transformer_engine/pytorch/module
NVTE_BACKWARD_OVERRIDE=dequantized python benchmarks/linear/benchmark_grouped_linear.py --recipe mxfp8 --fwd-only
       m     k     n recipe  num_gemms  grouped_fwd_time_ms
0  16384  7168  2048  mxfp8          4             0.282720
1  32768  7168  2048  mxfp8          4             0.526736
2  65536  7168  2048  mxfp8          4             0.982166
3  98304  7168  2048  mxfp8          4             1.451485
0  16384  7168  2048  mxfp8          8             0.313753
1  32768  7168  2048  mxfp8          8             0.551043
2  65536  7168  2048  mxfp8          8             1.040773
3  98304  7168  2048  mxfp8          8             1.527951
```

File: tests/cpp/operator/test_dequantize_nvfp4.cu [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    auto [atol, rtol] = getTolerances(otype);
+    if (num_elems > 0) {
+        compareResults("output_nvfp4_swizzled", output_swizzled,
+                       ref_output.get(), true, atol, rtol);
+    }
+}
+
+std::vector<std::pair<size_t, size_t>> nvfp4_tensor_dims = {
```
Comment (zhongbozhu) [inline tests/cpp/operator/test_dequantize_nvfp4.cu:215]: There is one edge case: 

For MXFP8, When the input shape is like 64x64, it will produce scaling factor shape 64x2, but then zero padded to 128x4. We should be able to inject some very large random values in the padded region during malloc (because we don't use torch.zeros to malloc but torch.empty), and detect whether dequantize results is affected. If things work as expected, this line will be triggered https://github.com/NVIDIA/TransformerEngine/blob/a073ad5b3ff5c1bc00d9e98669deabe901542aad/transformer_engine/common/cast/mxfp8/quantize_mxfp8.cuh#L759  and the dequantize numerics won't be affected.

For NVFP4, I think we optimize for GEMM (or swizzle fusion) is actually not enabled, same for the zero-out edge case handling logic? https://github.com/NVIDIA/TransformerEngine/blob/a073ad5b3ff5c1bc00d9e98669deabe901542aad/transformer_engine/common/cast/nvfp4/quantize_transpose_nvfp4.cuh#L1189 So there shouldn't be any unswizzle logic needed here?

Comment (zhongbozhu) [reply tests/cpp/operator/test_dequantize_nvfp4.cu:215]: For NVFP4, I believe currently only device-init grouped quantize with RHT has the swizzle fusion feature, so the scaling factor zero-out is the job of the dedicated swizzle kernel. So if we dequantize + unswizzle for NVFP4, the unswizzle logic might not be correct.

File: tests/cpp/operator/test_dequantize_nvfp4.cu [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    quantized_swizzled.set_amax(0.0f);
+    setRandomScale(&quantized_swizzled);
+    quantized_swizzled.set_with_gemm_swizzled_scales(true);
+
+    // Copy amax and scale from compact to swizzled before FP4 data,
+    // since from_cpu() uploads all CPU buffers (including zero-init data).
+    quantized_compact.to_cpu();
+    quantized_swizzled.set_amax(quantized_compact.amax());
+    quantized_swizzled.set_scale(quantized_compact.scale());
+    quantized_swizzled.from_cpu();
```
[the remark this answers is not in the document]
Comment (YigongQin) [reply tests/cpp/operator/test_dequantize_nvfp4.cu]: This redundant line has been removed

File: tests/cpp/operator/test_dequantize_nvfp4.cu (hunk 1 above)
Comment (YigongQin) [reply tests/cpp/operator/test_dequantize_nvfp4.cu:215]: For both MXFP8 and NVFP4, the unit test logic is: 1. generate compact scales (or from quantization); 2. call `nvte_swizzle_scaling_factors` to swizzle compact scales; 3. compare results of `nvte_dequantize` with compact scales and swizzled scales. Quantize with swizzle fusion is never enabled for both MXFP8 and NVFP4

File: transformer_engine/pytorch/module/linear.py [hunk 3]
```diff
@@ -1710,13 +1710,6 @@ def _get_quantizers(self, fp8_output, fp8_grad, is_grad_enabled):
                 grad_output_quantizer.optimize_for_gemm = True
             if fp8_grad:
                 grad_input_quantizer = self.quantizers["scaling_bwd"][FP8BwdTensorIdx.GRAD_INPUT1]
-        fp8_recipe = FP8GlobalStateManager.get_fp8_recipe()
-        if fp8_recipe.backward_override == "dequantized" and (
-            fp8_recipe.mxfp8() or fp8_recipe.nvfp4()
-        ):
-            input_quantizer.optimize_for_gemm = False
-            if grad_output_quantizer is not None:
-                grad_output_quantizer.optimize_for_gemm = False
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/linear.py:1751]: I'm of two minds about this:

- Logically, GEMM-optimized data is not guaranteed to support anything except GEMMs. Even if MXFP8 and NVFP4 dequant happens to support them, these are custom optimizations. Future recipes can not be expected to support dequantizing GEMM-optimzied data by default.
- It's a little pedantic to have edge-case logic that won't be triggered by any of our existing use-cases. Given how subtle this is, I worry about it becoming stale and distracting.

I think for now, this change is fine. However, if we encounter problems in a future recipe, we should reimplement it properly:
```python
# LOGICALLY WRONG!
# Fails if we add a new recipe
if recipe.backward_override == "dequantized" and recipe.future_recipe():
    input_quantizer.optimize_for_gemm = False

# LOGICALLY RIGHT!
# Automatically handles new recipes
if recipe.backward_override == "dequantized" and not (
    recipe.float8_per_tensor_scaling()
    or recipe.float8_block_scaling()
    or recipe.mxfp8()
    or recipe.nvfp4()
):
    input_quantizer.optimize_for_gemm = False
```

CC @ptrendx @ksivaman @zhongbozhu

File: tests/cpp/test_common.h [hunk 4]
```diff
@@ -286,6 +286,16 @@ class Tensor {
     tensor_.set_amax(nullptr, DType::kFloat32, tensor_.defaultShape);
   }
 
+  void set_amax(float amax_val) {
+    if (!amax_cpu_data_) {
+      amax_cpu_data_ = std::make_shared<float>(0);
+      float* amax_gpu = nullptr;
+      cudaMalloc((void**)&amax_gpu, sizeof(float));  // NOLINT(*)
```
Comment (timmoon10) [inline tests/cpp/test_common.h]: It's a code smell that amax needs this special allocation logic that different from all the other data fields (`data`, `columnwise_data`, `scale_inv`, etc).

The right fix is to allocate in the `Tensor` constructor, along with all the other GPU buffers:
https://github.com/NVIDIA/TransformerEngine/blob/0c2e7b09c6d33109803ea089fbf80421a326e0a4/tests/cpp/test_common.cu#L395

While checking this, I've also found that our current implementation leaks the scale and amax buffers. We should deallocate them in the `Tensor` destructor, along with all the other GPU buffers: https://github.com/NVIDIA/TransformerEngine/blob/0c2e7b09c6d33109803ea089fbf80421a326e0a4/tests/cpp/test_common.h#L155-L157

Comment (YigongQin) [reply tests/cpp/test_common.h]: adedd amax allocation in constructor and amax/scale deallocation in destructor

File: tests/cpp/test_common.h [hunk 5]
```diff
@@ -286,6 +294,11 @@ class Tensor {
     tensor_.set_amax(nullptr, DType::kFloat32, tensor_.defaultShape);
   }
 
+  void set_amax(float amax_val) {
+    NVTE_CHECK(amax_cpu_data_);
+    *amax_cpu_data_ = amax_val;
```
Comment (timmoon10) [inline tests/cpp/test_common.h]: We should make sure that the GPU is also updated, similar to `set_scale` and `set_scale_inv`.
```suggestion
    *amax_cpu_data_ = amax_val;
    from_cpu();
```

File: tests/cpp/test_common.cu [hunk 6]
```diff
@@ -396,6 +396,13 @@ Tensor::Tensor(const std::string& name,
         cudaMemset(scale, 0, sizeof(float));
         scale_cpu_data_ = std::make_shared<float>(0);
         tensor_.set_scale(scale, DType::kFloat32, std::vector<size_t>{1});
+
+        // Allocate amax alongside other GPU buffers
```
Comment (timmoon10) [inline tests/cpp/test_common.cu]: This is a very Claude-y comment. It's helpful for keeping track of its changes, but it provides no value in the final code. It's obvious we're allocating amax alongside scale, why are we mentioning it specially?
```suggestion
```

The problem with these comments is that the code just becomes littered with historical comments about why each line was added, and it no longer becomes a coherent thing that stands by itself. If I want to learn history I can look at Git, but the really important thing is how the code works _today_.

**Review by timmoon10 — DISMISSED** [review]
LGTM, pending CI. These kernels will be very useful.

**Review by timmoon10 — APPROVED** [review]
